Cargo.toml

```diff
@@ -41,11 +41,17 @@ path = "tests/integration/lib.rs"
 name = "e2e"
 path = "tests/e2e/mod.rs"
 
-# CI-based builds use full release optimization.  See scripts/environment/release-flags.sh.
-# This results in roughly a 5% reduction in performance when compiling locally vs when
-# compiled via the CI pipeline.
 [profile.release]
-debug = false # Do not include debug symbols in the executable.
+lto = true
```

Comment:
Is "fat" worthwhile vs "thin"? Technically there's a few variants here.
- `false` (default) will opt-out of LTO when `codegen-units=1` or `opt-level = 0`, otherwise it does crate local thin LTO (slightly different than "thin").
- `"thin"` => Thin LTO
- `"fat"` / `true` => Full LTO (much slower, barely any notable gain)
- `"off"` => Disable LTO

In my testing with build size and time, I didn't see much value from building Vector with `"fat"` / `true`. It was notably slower than `false` / `"thin"` due to single thread vs multi-thread CPU usage at link time.

From what I've read of those that profile between thin and fat/full LTO there's barely a statistical difference in improvement to warrant the excess build time.

Reply:
Good thought. Once this is merged, we can benchmark against `thin` and see what the difference looks like for `fat` vs. `thin`. I believe we saw a 5% difference in throughput for `fat` vs `false`.

Reply:
@polarathene There's a very big difference in how crates are handled between the two: https://doc.rust-lang.org/cargo/reference/profiles.html#lto

Reply:
I found [this blogpost](https://blog.rust-lang.org/inside-rust/2020/06/29/lto-improvements.html) that explains the difference for `lto = false` (thin local LTO) vs `lto ="thin"`, where `false` does thin LTO over each crates codegen units.

> I believe we saw a 5% difference in throughput for `fat` vs `false`.

That's interesting, would be good to know how that compes to `fat` vs `thin` 👍 

Was the `false` comparison done with `codegen-units = 1`? Because as I mentioned earlier, that would opt-out of LTO and be equivalent to `lto = "false"`.

https://blog.llvm.org/2016/06/thinlto-scalable-and-incremental-lto.html

> _ThinLTO already performs well compared to LTO, in many cases matching the performance improvement._
> _In a few cases ThinLTO even outperforms full LTO, most likely because the higher scalability of ThinLTO allows using a more aggressive backend optimization pipeline (similar to that of a non-LTO build)._

https://convolv.es/guides/lto/

> _our recurring dataset shows a 0.23% run-time delta between LLVM’s basic and parallel modes_
> _LLVM’s implementation of parallel LTO achieves nearly all of the run-time performance improvement as seen with basic LTO:_
> - _basic LTO reached a 2.86% improvement over non-LTO_
> - _while parallel LTO achieved a 2.63% improvement over non-LTO._

Not sure how much it differs between GCC vs Clang as the linker driver. I think with Clang you need to have a [compatible version to the Rust toolchain used](https://doc.rust-lang.org/rustc/linker-plugin-lto.html#toolchain-compatibility) (_at least if doing cross-language LTO_) which may not work as well with the CI image builds using `cross` docker images (Ubuntu)? 🤷‍♂️ (_[this comment](https://www.reddit.com/r/rust/comments/1bbcdzs/comment/ku8z32m/) seems to claim a 20% runtime perf improvement by changing linker from GCC to Clang_)